<a href="https://colab.research.google.com/github/preetamjumech/No-sql-MongoDB/blob/main/Preetam_Saha_C22018_Midterm_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Additional Material https://github.com/giffy/MongoDB_PyMongo_Tutorial

#Install MongoDB

In [252]:
!apt install mongodb > /dev/null
#!apt install mongodb 

In [253]:
#Local Server Need Not be started for Remote Operations
#!service mongodb start

In [254]:
from pymongo import MongoClient
import pandas as pd
import dateutil

## Test MongoClient with Remote Server

In [255]:
# The Password is visible in this cell
# If you wish to hide this then use the next cell
#
rhostName = 'rs0/n1-c2-mongodb-clevercloud-customers.services.clever-cloud.com:27017,n2-c2-mongodb-clevercloud-customers.services.clever-cloud.com:27017'   # place your own credentials here
ruserName = 'uwug2aas6fid9qyr5bit'      # place your own credentials here
rpassWord = 'yeNDLftiFKmQJe4R6GMz'      # place your own credentials here
rdbName =  'begat3qdocpcl0t'         # place your own credentials here
rMongoURI = 'mongodb://uwug2aas6fid9qyr5bit:yeNDLftiFKmQJe4R6GMz@n1-c2-mongodb-clevercloud-customers.services.clever-cloud.com:27017,n2-c2-mongodb-clevercloud-customers.services.clever-cloud.com:27017/begat3qdocpcl0t?replicaSet=rs0'         # place your own credentials here

In [256]:
mongoURI = rMongoURI                                   # for remote server
#mongoURI = "mongodb://localhost:27017/"                 # for local server
client = MongoClient(mongoURI)
client.list_database_names() # ['admin', 'local']

['begat3qdocpcl0t']

In [257]:
#db = client.PraxisDB
db = client.begat3qdocpcl0t
db.list_collection_names()

['restaurants', 'une']

In [258]:
table = db.restaurants
table.count_documents({})

3772

#MongoDB Exercise <br>
Exercise : https://docs.google.com/document/d/1q0PyJL6EgMG1V5G8mmPre8Ypq9N9dX-PRLixAHiphd8/edit?usp=sharing <br>

Data : restaurant.json https://drive.google.com/file/d/14sYJM2KwIpDRsAAk6YkapMTwLnLuBcb_/view?usp=sharing <br>

For more detailed examples, see https://colab.research.google.com/github/Giffy/MongoDB_PyMongo_Tutorial/blob/master/1_1_First_steps_with_MongoDB_PyMongo.ipynb

## 2.1Load Data

## 2.2 Write a short explanation of what the data is all about, what fields are being used

**data link https://data.world/badosa/uneces-country-overview
Data is about population of particular country of different years. 
Here the important features are:

1.area_square_kilometres

2.total_population

3.population_density_pers_per_sq_km

4.population_aged_0_14_male

5.population_aged_0_14_female

6.total_population_male

7.total_population_female

8.life_expectancy_at_birth_women

9.life_expectancy_at_birth_men

10.total_employment_growth_rate

11.unemployment_rate

12.persons_killed_in_road_accidents

13.persons_injured_in_road_accidents

14.total_length_of_motorways_km

15.total_length_of_railway_lines_km**

In [259]:
db.drop_collection('une')

{'$clusterTime': {'clusterTime': Timestamp(1656007831, 1),
  'signature': {'hash': b'\x96\xe9>\xe2\x9f9j\x9bKl\xd7A\xe1\xc3\xe5\x18\xd8.\xb2\xda',
   'keyId': 7106106537503883265}},
 'nIndexesWas': 1,
 'ns': 'begat3qdocpcl0t.une',
 'ok': 1.0,
 'operationTime': Timestamp(1656007831, 1)}

In [260]:
# the restaurants.json file is available in my GDrive
# Local Server
#!mongoimport --db PraxisDB --collection restaurants --drop --file restaurants.json
# Remote Server
# mongoimport --host "hostname" --port "port" --db "databasename" --collection "collectionName" --file "filePath"
# the python variables need to be placed in {}
#!mongoimport --host "{rhostName}" --port "27017" --db "{rdbName}" --collection "restaurants"  --drop --file "restaurants.json" -u "{ruserName}" -p "{rpassWord}" --- THIS DOES NOT WORK
!mongoimport --uri "{rMongoURI}"  --collection "une"  --drop --file "/content/unece.json" --jsonArray

2022-06-23T18:10:31.609+0000	connected to: localhost
2022-06-23T18:10:31.625+0000	dropping: begat3qdocpcl0t.une
2022-06-23T18:10:32.116+0000	imported 884 documents


In [261]:
db = client.begat3qdocpcl0t
db.list_collection_names()

['restaurants', 'une']

In [262]:
table = db.une
table.count_documents({})

884

##Data Access & Manipulation

In [263]:
#mongoURI = "mongodb://localhost:27017/"    # local
mongoURI = rMongoURI                        # remote
client = MongoClient(mongoURI)
client.list_database_names() 

['begat3qdocpcl0t']

In [264]:
#db = client.PraxisDB
db = client.begat3qdocpcl0t
db.list_collection_names()

['restaurants', 'une']

In [265]:
table = db.une
table.count_documents({})

884

# 32 Diverse Queries

## 2.3 5 different queries specifying the selection conditions and projection conditions

In [266]:
table = db.une
c1 = table.find({},{"Country":1,"Year" : 1,"_id":0,'Area (square kilometres)':1,'Total population':1,
                    "Persons killed in road accidents":1,"Total length of railway lines (km)":1,
                    "Import of goods and services, per cent of GDP":1,
                    'Population aged 0-14, male':1,'Population aged 0-14, female':1}).limit(1000)
pr=[]
for crnt in c1:
    print(crnt)
    country=crnt["Country"]
    year=crnt["Year"]
    area=crnt['Area (square kilometres)']
    population=crnt['Total population']
    person_killed=crnt["Persons killed in road accidents"]
    length_railway=crnt["Total length of railway lines (km)"]
    goods_services=crnt["Import of goods and services, per cent of GDP"]
    male_pop=crnt['Population aged 0-14, male']
    female_pop=crnt["Population aged 0-14, female"]
    info=(country,year,area,population,person_killed,length_railway,goods_services,male_pop,female_pop)
    pr.append(info)

{'Country': 'Albania', 'Year': '2000', 'Area (square kilometres)': 28748, 'Total population': 3401198, 'Population aged 0-14, male': 564851, 'Population aged 0-14, female': 529820, 'Import of goods and services, per cent of GDP': 43.33879, 'Persons killed in road accidents': 280, 'Total length of railway lines (km)': 400}
{'Country': 'Albania', 'Year': '2001', 'Area (square kilometres)': 28748, 'Total population': 3073734, 'Population aged 0-14, male': 460732, 'Population aged 0-14, female': 436652, 'Import of goods and services, per cent of GDP': 45.15763, 'Persons killed in road accidents': 297, 'Total length of railway lines (km)': 409}
{'Country': 'Albania', 'Year': '2002', 'Area (square kilometres)': 28748, 'Total population': 3093465, 'Population aged 0-14, male': 452373, 'Population aged 0-14, female': 427711, 'Import of goods and services, per cent of GDP': 47.57943, 'Persons killed in road accidents': 250, 'Total length of railway lines (km)': 428}
{'Country': 'Albania', 'Year

In [267]:
c1 = db.une.find({},{"_id":0}).limit(10)
for crnt in c1:
    print(crnt)

{'Country': 'Albania', 'Year': '2000', 'Area (square kilometres)': 28748, 'Total population': 3401198, 'Population density, pers. per sq. km': 118.31077, 'Population aged 0-14, male': 564851, 'Population aged 0-14, female': 529820, 'Population aged 15-64, male': 1029446, 'Population aged 15-64, female': 1086946, 'Population aged 64+, male': 82588, 'Population aged 64+, female': 107547, 'Total population, male (%)': 49.30278, 'Total population, female (%)': 50.69722, 'Life expectancy at birth, women': 78.11, 'Life expectancy at birth, men': 72.07, 'Life expectancy at age 65, women': 17.16, 'Life expectancy at age 65, men': 13.52, 'Total fertility rate': 2.4, 'Adolescent fertility rate': 15.24086, 'Mean age of women at birth of first child': None, 'Computer use, 16-24, male': None, 'Computer use, 16-24, female': None, 'Computer use, 25-54, male': None, 'Computer use, 25-54, female': None, 'Computer use, 55-74, male': None, 'Computer use, 55-74, female': None, 'Women in the Labour Force, 

**Projection**

In [268]:
c1 = db.une.find({},{"Area (square kilometres)" : 1,"Total population":1,"_id" :0}).limit(10)
for crnt in c1:
    print(crnt)

{'Area (square kilometres)': 28748, 'Total population': 3401198}
{'Area (square kilometres)': 28748, 'Total population': 3073734}
{'Area (square kilometres)': 28748, 'Total population': 3093465}
{'Area (square kilometres)': 28748, 'Total population': 3111162}
{'Area (square kilometres)': 28748, 'Total population': 3127264}
{'Area (square kilometres)': 28748, 'Total population': 3142065}
{'Area (square kilometres)': 28748, 'Total population': 3146813}
{'Area (square kilometres)': 28748, 'Total population': 3161337}
{'Area (square kilometres)': 28748, 'Total population': 3181950}
{'Area (square kilometres)': 28748, 'Total population': 3194417}


In [269]:
c1 = db.une.find({},{'Population aged 0-14, male':1,"_id":0}).limit(10)
for crnt in c1:
    print(crnt)

{'Population aged 0-14, male': 564851}
{'Population aged 0-14, male': 460732}
{'Population aged 0-14, male': 452373}
{'Population aged 0-14, male': 442939}
{'Population aged 0-14, male': 433178}
{'Population aged 0-14, male': 422474}
{'Population aged 0-14, male': 412338}
{'Population aged 0-14, male': 403781}
{'Population aged 0-14, male': 396600}
{'Population aged 0-14, male': 388901}


In [270]:
c1 = db.une.find({},{'Unemployment rate':1,"_id":0}).limit(10)
for crnt in c1:
    print(crnt)

{'Unemployment rate': 16.8}
{'Unemployment rate': 16.4}
{'Unemployment rate': 15.8}
{'Unemployment rate': 15}
{'Unemployment rate': 14.4}
{'Unemployment rate': 14.1}
{'Unemployment rate': 13.8}
{'Unemployment rate': 13.36596}
{'Unemployment rate': 13.05188}
{'Unemployment rate': 13.76226}


In [271]:
c1 = db.une.find({},{'Import of goods and services, per cent of GDP':1,"_id":0}).limit(10)
for crnt in c1:
    print(crnt)

{'Import of goods and services, per cent of GDP': 43.33879}
{'Import of goods and services, per cent of GDP': 45.15763}
{'Import of goods and services, per cent of GDP': 47.57943}
{'Import of goods and services, per cent of GDP': 46.17322}
{'Import of goods and services, per cent of GDP': 44.82965}
{'Import of goods and services, per cent of GDP': 47.85934}
{'Import of goods and services, per cent of GDP': 49.05747}
{'Import of goods and services, per cent of GDP': 54.95422}
{'Import of goods and services, per cent of GDP': 56.43098}
{'Import of goods and services, per cent of GDP': 53.80403}


In [272]:
c1 = db.une.find({},{'Total length of railway lines (km)':1,"_id":0}).limit(10)
for crnt in c1:
    print(crnt)

{'Total length of railway lines (km)': 400}
{'Total length of railway lines (km)': 409}
{'Total length of railway lines (km)': 428}
{'Total length of railway lines (km)': 428}
{'Total length of railway lines (km)': 428}
{'Total length of railway lines (km)': None}
{'Total length of railway lines (km)': None}
{'Total length of railway lines (km)': None}
{'Total length of railway lines (km)': None}
{'Total length of railway lines (km)': None}


## Selection

In [273]:
c1 = db.une.find({"Persons killed in road accidents":{"$gt" : 200}},{"_id":0}).limit(5)
for crnt in c1:
    print(crnt)

{'Country': 'Albania', 'Year': '2000', 'Area (square kilometres)': 28748, 'Total population': 3401198, 'Population density, pers. per sq. km': 118.31077, 'Population aged 0-14, male': 564851, 'Population aged 0-14, female': 529820, 'Population aged 15-64, male': 1029446, 'Population aged 15-64, female': 1086946, 'Population aged 64+, male': 82588, 'Population aged 64+, female': 107547, 'Total population, male (%)': 49.30278, 'Total population, female (%)': 50.69722, 'Life expectancy at birth, women': 78.11, 'Life expectancy at birth, men': 72.07, 'Life expectancy at age 65, women': 17.16, 'Life expectancy at age 65, men': 13.52, 'Total fertility rate': 2.4, 'Adolescent fertility rate': 15.24086, 'Mean age of women at birth of first child': None, 'Computer use, 16-24, male': None, 'Computer use, 16-24, female': None, 'Computer use, 25-54, male': None, 'Computer use, 25-54, female': None, 'Computer use, 55-74, male': None, 'Computer use, 55-74, female': None, 'Women in the Labour Force, 

In [274]:
c1 = db.une.find({"Area (square kilometres)":{"$gt" : 28000}},{"_id":0}).limit(5)
for crnt in c1:
    print(crnt)

{'Country': 'Albania', 'Year': '2000', 'Area (square kilometres)': 28748, 'Total population': 3401198, 'Population density, pers. per sq. km': 118.31077, 'Population aged 0-14, male': 564851, 'Population aged 0-14, female': 529820, 'Population aged 15-64, male': 1029446, 'Population aged 15-64, female': 1086946, 'Population aged 64+, male': 82588, 'Population aged 64+, female': 107547, 'Total population, male (%)': 49.30278, 'Total population, female (%)': 50.69722, 'Life expectancy at birth, women': 78.11, 'Life expectancy at birth, men': 72.07, 'Life expectancy at age 65, women': 17.16, 'Life expectancy at age 65, men': 13.52, 'Total fertility rate': 2.4, 'Adolescent fertility rate': 15.24086, 'Mean age of women at birth of first child': None, 'Computer use, 16-24, male': None, 'Computer use, 16-24, female': None, 'Computer use, 25-54, male': None, 'Computer use, 25-54, female': None, 'Computer use, 55-74, male': None, 'Computer use, 55-74, female': None, 'Women in the Labour Force, 

In [275]:
c1 = db.une.find({"Year": '2000'},{"_id":0}).limit(5)
for crnt in c1:
    print(crnt)

{'Country': 'Albania', 'Year': '2000', 'Area (square kilometres)': 28748, 'Total population': 3401198, 'Population density, pers. per sq. km': 118.31077, 'Population aged 0-14, male': 564851, 'Population aged 0-14, female': 529820, 'Population aged 15-64, male': 1029446, 'Population aged 15-64, female': 1086946, 'Population aged 64+, male': 82588, 'Population aged 64+, female': 107547, 'Total population, male (%)': 49.30278, 'Total population, female (%)': 50.69722, 'Life expectancy at birth, women': 78.11, 'Life expectancy at birth, men': 72.07, 'Life expectancy at age 65, women': 17.16, 'Life expectancy at age 65, men': 13.52, 'Total fertility rate': 2.4, 'Adolescent fertility rate': 15.24086, 'Mean age of women at birth of first child': None, 'Computer use, 16-24, male': None, 'Computer use, 16-24, female': None, 'Computer use, 25-54, male': None, 'Computer use, 25-54, female': None, 'Computer use, 55-74, male': None, 'Computer use, 55-74, female': None, 'Women in the Labour Force, 

In [276]:
c1 = db.une.find({"Female tertiary students, percent of total": {"$gt" : 55}},{"_id":0}).limit(6)
for crnt in c1:
    print(crnt)

{'Country': 'Albania', 'Year': '2000', 'Area (square kilometres)': 28748, 'Total population': 3401198, 'Population density, pers. per sq. km': 118.31077, 'Population aged 0-14, male': 564851, 'Population aged 0-14, female': 529820, 'Population aged 15-64, male': 1029446, 'Population aged 15-64, female': 1086946, 'Population aged 64+, male': 82588, 'Population aged 64+, female': 107547, 'Total population, male (%)': 49.30278, 'Total population, female (%)': 50.69722, 'Life expectancy at birth, women': 78.11, 'Life expectancy at birth, men': 72.07, 'Life expectancy at age 65, women': 17.16, 'Life expectancy at age 65, men': 13.52, 'Total fertility rate': 2.4, 'Adolescent fertility rate': 15.24086, 'Mean age of women at birth of first child': None, 'Computer use, 16-24, male': None, 'Computer use, 16-24, female': None, 'Computer use, 25-54, male': None, 'Computer use, 25-54, female': None, 'Computer use, 55-74, male': None, 'Computer use, 55-74, female': None, 'Women in the Labour Force, 

In [277]:
c1 = db.une.find({"Total employment, growth rate":{"$gt" : 0}},{"_id":0}).limit(6)
for crnt in c1:
    print(crnt)

{'Country': 'Albania', 'Year': '2000', 'Area (square kilometres)': 28748, 'Total population': 3401198, 'Population density, pers. per sq. km': 118.31077, 'Population aged 0-14, male': 564851, 'Population aged 0-14, female': 529820, 'Population aged 15-64, male': 1029446, 'Population aged 15-64, female': 1086946, 'Population aged 64+, male': 82588, 'Population aged 64+, female': 107547, 'Total population, male (%)': 49.30278, 'Total population, female (%)': 50.69722, 'Life expectancy at birth, women': 78.11, 'Life expectancy at birth, men': 72.07, 'Life expectancy at age 65, women': 17.16, 'Life expectancy at age 65, men': 13.52, 'Total fertility rate': 2.4, 'Adolescent fertility rate': 15.24086, 'Mean age of women at birth of first child': None, 'Computer use, 16-24, male': None, 'Computer use, 16-24, female': None, 'Computer use, 25-54, male': None, 'Computer use, 25-54, female': None, 'Computer use, 55-74, male': None, 'Computer use, 55-74, female': None, 'Women in the Labour Force, 

2.4 Install Spark in the local VM and pull some fields from MongoDB to Spark dataframes

**Spark with MONGO**

In [278]:
from pymongo import MongoClient
import pandas as pd
import os

In [279]:
!apt install mongodb > /dev/null
!service mongodb start



 * Starting database mongodb
   ...done.


In [280]:
mongoURI = "mongodb://localhost:27017/"
client = MongoClient(mongoURI)
client.list_database_names() # ['admin', 'local']

['admin', 'config', 'local', 'people', 'sparkDB']

In [281]:
db = client.local
#db = client.sparkDB
#db = client.people
db.list_collection_names()

['startup_log']

In [282]:
table = db.startup_log
table.find_one()

{'_id': '6e31e4231155-1656001049045',
 'buildinfo': {'allocator': 'tcmalloc',
  'bits': 64,
  'buildEnvironment': {'cc': 'cc: cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0',
   'ccflags': '-fno-omit-frame-pointer -fno-strict-aliasing -ggdb -pthread -Wall -Wsign-compare -Wno-unknown-pragmas -Wno-error=c++1z-compat -Wno-error=noexcept-type -Wno-error=format-truncation -Wno-error=int-in-bool-context -Winvalid-pch -Werror -O2 -Wno-unused-local-typedefs -Wno-unused-function -Wno-deprecated-declarations -Wno-unused-const-variable -Wno-unused-but-set-variable -Wno-missing-braces -fstack-protector-strong -fno-builtin-memcmp',
   'cxx': 'g++: g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0',
   'cxxflags': '-g -O2 -fdebug-prefix-map=/build/mongodb-FmqHhf/mongodb-3.6.3=. -fstack-protector-strong -Wformat -Werror=format-security -Woverloaded-virtual -Wno-maybe-uninitialized -std=c++14',
   'distarch': 'x86_64',
   'distmod': '',
   'linkflags': '-Wl,-Bsymbolic-functions -Wl,-z,relro -pthread -Wl,-z,now -rdynami

In [283]:
!pip3 install -q pyspark

In [284]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .master("local[*]") \
    .config("spark.mongodb.input.uri", "mongodb://localhost:27017/sparkDB.coll01") \
    .config("spark.mongodb.output.uri", "mongodb://localhost:27017/sparkDB.coll01") \
    .config('spark.jars.packages', 'org.mongodb.spark:mongo-spark-connector_2.12:3.0.1')\
    .getOrCreate()

In [285]:
sc = spark.sparkContext
sc

<SparkContext master=local[*] appName=pyspark-shell>

In [286]:
df=pd.DataFrame(pr)

In [287]:
df.head()

0     1        2          3      4      5         6         7  \
0  Albania  2000  28748.0  3401198.0  280.0  400.0  43.33879  564851.0   
1  Albania  2001  28748.0  3073734.0  297.0  409.0  45.15763  460732.0   
2  Albania  2002  28748.0  3093465.0  250.0  428.0  47.57943  452373.0   
3  Albania  2003  28748.0  3111162.0  264.0  428.0  46.17322  442939.0   
4  Albania  2004  28748.0  3127264.0  315.0  428.0  44.82965  433178.0   

          8  
0  529820.0  
1  436652.0  
2  427711.0  
3  417968.0  
4  407988.0

In [288]:
df.columns=["country","year","Area","Population","person_killed","length_railway","goods_services","male_pop","female_pop"]

In [289]:
df.head(6)

country  year     Area  Population  person_killed  length_railway  \
0  Albania  2000  28748.0   3401198.0          280.0           400.0   
1  Albania  2001  28748.0   3073734.0          297.0           409.0   
2  Albania  2002  28748.0   3093465.0          250.0           428.0   
3  Albania  2003  28748.0   3111162.0          264.0           428.0   
4  Albania  2004  28748.0   3127264.0          315.0           428.0   
5  Albania  2005  28748.0   3142065.0          307.0             NaN   

   goods_services  male_pop  female_pop  
0        43.33879  564851.0    529820.0  
1        45.15763  460732.0    436652.0  
2        47.57943  452373.0    427711.0  
3        46.17322  442939.0    417968.0  
4        44.82965  433178.0    407988.0  
5        47.85934  422474.0    397035.0

In [290]:
df1=df.dropna()

In [291]:
people = spark.createDataFrame(df1)

In [292]:
people.show()

+----------+----+-------+----------+-------------+--------------+--------------+---------+----------+
|   country|year|   Area|Population|person_killed|length_railway|goods_services| male_pop|female_pop|
+----------+----+-------+----------+-------------+--------------+--------------+---------+----------+
|   Albania|2000|28748.0| 3401198.0|        280.0|         400.0|      43.33879| 564851.0|  529820.0|
|   Albania|2001|28748.0| 3073734.0|        297.0|         409.0|      45.15763| 460732.0|  436652.0|
|   Albania|2002|28748.0| 3093465.0|        250.0|         428.0|      47.57943| 452373.0|  427711.0|
|   Albania|2003|28748.0| 3111162.0|        264.0|         428.0|      46.17322| 442939.0|  417968.0|
|   Albania|2004|28748.0| 3127264.0|        315.0|         428.0|      44.82965| 433178.0|  407988.0|
|   Armenia|2000|29743.0| 3221105.0|        214.0|         842.9|      50.54329| 406848.0|  382516.0|
|   Armenia|2001|29743.0| 3211815.0|        237.0|         842.9|      46.14881| 4

In [293]:
people

DataFrame[country: string, year: string, Area: double, Population: double, person_killed: double, length_railway: double, goods_services: double, male_pop: double, female_pop: double]

In [294]:
people.count()

466

In [295]:
people.write.format("mongo").mode("append").save()

In [296]:
people.write.format("mongo").mode("append").option("database","people").option("collection", "une1").save()

In [297]:
client.list_database_names()

['admin', 'config', 'local', 'people', 'sparkDB']

In [298]:
db = client.people
list(db.list_collections())

[{'idIndex': {'key': {'_id': 1},
   'name': '_id_',
   'ns': 'people.airlines1',
   'v': 2},
  'info': {'readOnly': False,
   'uuid': Binary(b'\xb3>"B\x96\xadC\x82\xa8"=\x17a;\xa6\x01', 4)},
  'name': 'airlines1',
  'options': {},
  'type': 'collection'},
 {'idIndex': {'key': {'_id': 1}, 'name': '_id_', 'ns': 'people.une1', 'v': 2},
  'info': {'readOnly': False,
   'uuid': Binary(b'\xdaE\xd2\xf3\x9d\x81M|\x97rrc.mb6', 4)},
  'name': 'une1',
  'options': {},
  'type': 'collection'}]

In [299]:
db2 = client.sparkDB
list(db2.list_collections())

[{'idIndex': {'key': {'_id': 1},
   'name': '_id_',
   'ns': 'sparkDB.coll01',
   'v': 2},
  'info': {'readOnly': False,
   'uuid': Binary(b'\xfe0\xad\x15g\xabH\xa5\x80\xec3L\xba\\\xa2\xb1', 4)},
  'name': 'coll01',
  'options': {},
  'type': 'collection'}]

In [300]:
kollection = db.une1
c = kollection.find({},{"_id":0})
for crnt in c:
    print(crnt)

{'country': 'Kyrgyzstan', 'year': '2011', 'Area': 199900.0, 'Population': 5514754.0, 'male_pop': 854192.0, 'female_pop': 820942.0}
{'country': 'Kyrgyzstan', 'year': '2012', 'Area': 199900.0, 'Population': 5607511.0, 'male_pop': 876533.0, 'female_pop': 841691.0}
{'country': 'Kyrgyzstan', 'year': '2013', 'Area': 199900.0, 'Population': 5719852.0, 'male_pop': 902193.0, 'female_pop': 865120.0}
{'country': 'Kyrgyzstan', 'year': '2014', 'Area': 199900.0, 'Population': 5835816.0, 'male_pop': 929639.0, 'female_pop': 890057.0}
{'country': 'Kyrgyzstan', 'year': '2015', 'Area': 199900.0, 'Population': 5957271.0, 'male_pop': 960742.0, 'female_pop': 918127.0}
{'country': 'Latvia', 'year': '2000', 'Area': 64589.0, 'Population': 2367549.5, 'male_pop': 213930.5, 'female_pop': 204477.5}
{'country': 'Latvia', 'year': '2001', 'Area': 64589.0, 'Population': 2337170.0, 'male_pop': 203524.5, 'female_pop': 194791.0}
{'country': 'Latvia', 'year': '2002', 'Area': 64589.0, 'Population': 2310173.0, 'male_pop': 1

In [301]:
pandasDF = pd.DataFrame(kollection.find({},{"_id":0}))
pandasDF


country  year       Area   Population    male_pop  female_pop  \
0        Kyrgyzstan  2011   199900.0    5514754.0    854192.0    820942.0   
1        Kyrgyzstan  2012   199900.0    5607511.0    876533.0    841691.0   
2        Kyrgyzstan  2013   199900.0    5719852.0    902193.0    865120.0   
3        Kyrgyzstan  2014   199900.0    5835816.0    929639.0    890057.0   
4        Kyrgyzstan  2015   199900.0    5957271.0    960742.0    918127.0   
...             ...   ...        ...          ...         ...         ...   
1712  United States  2007  9629090.0  301231207.0  31030581.0  29651034.0   
1713  United States  2008  9629090.0  304093966.0  31136239.0  29771145.0   
1714  United States  2010  9629090.0  309346863.0  31272117.0  29930297.0   
1715  United States  2009  9629090.0  306771529.0  31221676.0  29865905.0   
1716  United States  2013  9629090.0  316427395.0  31198756.0  29887133.0   

      person_killed  length_railway  goods_services  
0               NaN             NaN             NaN  
1               NaN             NaN             NaN  
2               NaN             NaN             NaN  
3               NaN             NaN             NaN  
4               NaN             NaN             NaN  
...             ...             ...             ...  
1712        41259.0        151986.0        16.46129  
1713        37423.0        151615.0        17.42693  
1714        32999.0        154014.0        15.80418  
1715        33808.0        151356.0        13.75436  
1716        32719.0        153266.0        16.63366  

[1717 rows x 9 columns]

In [315]:
sparkDF = spark.read.format("mongo").option("uri","mongodb://127.0.0.1/people.une1").load()
sparkDF.show(40)

+--------+----------+--------------------+----------+----------+--------------+--------------+--------+-------------+----+
|    Area|Population|                 _id|   country|female_pop|goods_services|length_railway|male_pop|person_killed|year|
+--------+----------+--------------------+----------+----------+--------------+--------------+--------+-------------+----+
|199900.0| 5514754.0|{62b4950a2294945e...|Kyrgyzstan|  820942.0|          null|          null|854192.0|         null|2011|
|199900.0| 5607511.0|{62b4950a2294945e...|Kyrgyzstan|  841691.0|          null|          null|876533.0|         null|2012|
|199900.0| 5719852.0|{62b4950a2294945e...|Kyrgyzstan|  865120.0|          null|          null|902193.0|         null|2013|
|199900.0| 5835816.0|{62b4950a2294945e...|Kyrgyzstan|  890057.0|          null|          null|929639.0|         null|2014|
|199900.0| 5957271.0|{62b4950a2294945e...|Kyrgyzstan|  918127.0|          null|          null|960742.0|         null|2015|
| 64589.0| 23675

## Statistical Analysis

In [303]:
sparkDF.count()

1717

In [304]:
sparkDF.select("country").distinct().count()

52

In [305]:
sparkDF.select("year").distinct().show()

+----+
|year|
+----+
|2012|
|2014|
|2013|
|2005|
|2000|
|2002|
|2009|
|2006|
|2004|
|2011|
|2008|
|2007|
|2015|
|2001|
|2010|
|2003|
+----+



In [306]:
sparkDF.groupby('country').agg({'female_pop': 'mean'}).show()

+--------------------+------------------+
|             country|   avg(female_pop)|
+--------------------+------------------+
|The former Yugosl...|189771.05952380953|
|              Sweden| 772670.7564102564|
|              Turkey|  9183799.15116279|
|             Germany| 5638966.384615385|
|              France| 5777765.121621622|
|              Greece| 799411.3181818182|
|            Slovakia| 438706.0232558139|
|             Belgium| 887264.8714285714|
|             Albania|391375.73076923075|
|             Finland|443730.77906976745|
|       United States|       2.9686011E7|
|             Belarus|        746366.875|
|               Malta| 32950.76666666667|
|          Tajikistan| 1288795.423076923|
|             Croatia|        327832.325|
|               Italy|  4005209.11627907|
|           Lithuania| 264258.9146341463|
|              Norway| 444583.4756097561|
|        Turkmenistan|          763242.7|
|               Spain| 3175650.717948718|
+--------------------+------------

In [307]:
sparkDF.groupby('country').agg({'male_pop': 'max'}).show()

+--------------------+-------------+
|             country|max(male_pop)|
+--------------------+-------------+
|The former Yugosl...|     233288.5|
|              Sweden|     854957.0|
|              Turkey|    9989599.0|
|             Germany|    6588158.0|
|              France|    6290810.0|
|              Greece|     870485.5|
|            Slovakia|     537886.5|
|             Belgium|     977159.0|
|             Albania|     564851.0|
|             Finland|     479585.5|
|       United States|  3.1272117E7|
|             Belarus|     972280.0|
|               Malta|      39851.0|
|          Tajikistan|    1434137.0|
|             Croatia|     368412.5|
|               Italy|    4330798.0|
|           Lithuania|     357545.5|
|              Norway|     477058.5|
|        Turkmenistan|     824644.0|
|               Spain|    3648803.5|
+--------------------+-------------+
only showing top 20 rows



In [308]:
sparkDF.groupby('country').agg({'female_pop': 'mean'}).show()

+--------------------+------------------+
|             country|   avg(female_pop)|
+--------------------+------------------+
|The former Yugosl...|189771.05952380953|
|              Sweden| 772670.7564102564|
|              Turkey|  9183799.15116279|
|             Germany| 5638966.384615385|
|              France| 5777765.121621622|
|              Greece| 799411.3181818182|
|            Slovakia| 438706.0232558139|
|             Belgium| 887264.8714285714|
|             Albania|391375.73076923075|
|             Finland|443730.77906976745|
|       United States|       2.9686011E7|
|             Belarus|        746366.875|
|               Malta| 32950.76666666667|
|          Tajikistan| 1288795.423076923|
|             Croatia|        327832.325|
|               Italy|  4005209.11627907|
|           Lithuania| 264258.9146341463|
|              Norway| 444583.4756097561|
|        Turkmenistan|          763242.7|
|               Spain| 3175650.717948718|
+--------------------+------------

In [309]:
sparkDF.groupby('country').agg({'female_pop': 'min'}).show()

+--------------------+---------------+
|             country|min(female_pop)|
+--------------------+---------------+
|The former Yugosl...|       167306.0|
|              Sweden|       750946.0|
|              Turkey|      9075309.0|
|             Germany|      5192584.5|
|              France|      5658569.0|
|              Greece|       764301.0|
|            Slovakia|       404214.5|
|             Belgium|       877662.0|
|             Albania|       250221.0|
|             Finland|       434267.5|
|       United States|    2.9415223E7|
|             Belarus|       682511.0|
|               Malta|        29779.0|
|          Tajikistan|      1258888.0|
|             Croatia|       303327.5|
|               Italy|      3945513.0|
|           Lithuania|       208592.5|
|              Norway|       437295.5|
|        Turkmenistan|       722232.0|
|               Spain|      2891521.5|
+--------------------+---------------+
only showing top 20 rows



In [310]:
sparkDF.groupby('country').agg({'area': 'sum'}).show()

+--------------------+-----------+
|             country|  sum(area)|
+--------------------+-----------+
|The former Yugosl...|  1079946.0|
|              Sweden|1.7548596E7|
|              Turkey|3.3317045E7|
|             Germany| 1.392417E7|
|              France|  2.04055E7|
|              Greece|  4354581.0|
|            Slovakia|  2107516.0|
|             Belgium|  1068480.0|
|             Albania|   747448.0|
|             Finland| 1.454045E7|
|       United States| 3.851636E8|
|             Belarus|  3321600.0|
|               Malta|     4740.0|
|          Tajikistan|  1860300.0|
|             Croatia|  2261680.0|
|               Italy|1.2957491E7|
|           Lithuania|  2677300.0|
|              Norway|1.3274119E7|
|        Turkmenistan|  4881000.0|
|               Spain|1.9733688E7|
+--------------------+-----------+
only showing top 20 rows



In [311]:
sparkDF.groupby('country').agg({'male_pop': 'mean'}).show()

+--------------------+------------------+
|             country|     avg(male_pop)|
+--------------------+------------------+
|The former Yugosl...| 202384.7142857143|
|              Sweden| 814488.2948717949|
|              Turkey| 9655687.430232558|
|             Germany| 5941185.551282051|
|              France| 6061432.635135135|
|              Greece| 838930.9090909091|
|            Slovakia|461253.29069767444|
|             Belgium|          927883.4|
|             Albania|418090.92307692306|
|             Finland| 463065.4651162791|
|       United States|    3.1066941075E7|
|             Belarus|       788954.6875|
|               Malta| 34861.23333333333|
|          Tajikistan|1348777.7692307692|
|             Croatia|       344881.7625|
|               Italy| 4235530.872093023|
|           Lithuania| 277892.3048780488|
|              Norway|  467530.756097561|
|        Turkmenistan|          781704.5|
|               Spain|3361724.9615384615|
+--------------------+------------

In [316]:
sparkDF.groupby('country').agg({'male_pop': 'max'}).show()

+--------------------+-------------+
|             country|max(male_pop)|
+--------------------+-------------+
|The former Yugosl...|     233288.5|
|              Sweden|     854957.0|
|              Turkey|    9989599.0|
|             Germany|    6588158.0|
|              France|    6290810.0|
|              Greece|     870485.5|
|            Slovakia|     537886.5|
|             Belgium|     977159.0|
|             Albania|     564851.0|
|             Finland|     479585.5|
|       United States|  3.1272117E7|
|             Belarus|     972280.0|
|               Malta|      39851.0|
|          Tajikistan|    1434137.0|
|             Croatia|     368412.5|
|               Italy|    4330798.0|
|           Lithuania|     357545.5|
|              Norway|     477058.5|
|        Turkmenistan|     824644.0|
|               Spain|    3648803.5|
+--------------------+-------------+
only showing top 20 rows



In [312]:
from pyspark.mllib.stat import Statistics 
from math import sqrt 
from pyspark.sql.functions import mean


sparkDF.select(mean("Population")).collect()

[Row(avg(Population)=26493909.18899243)]

In [313]:
from pyspark.sql.functions import sum


sparkDF.select(sum("Population")).collect()

[Row(sum(Population)=45490042077.5)]

 ## 2.6 Write a few words to explain your data analysis

#### 1.In Turkey, max male_population is 9989599

#### 2.In Germany average male_population is 5941185

#### 3.In Italy min female_population is 3945513

#### 4. There are 52 distinct countries

#### 5. Total Population of all the coutries is 45490042077